### ISSUE TITLE
Kafka Memory issue with large number of consumer threads consuming from same topic on windows machine

### ISSUE THREAD

**Issue by Sandesh-J** [confluentinc/confluent-kafka-dotnet#1353]
Description
===========

We have a Kafka consumer application( written in .NET CORE 2.1) which is consuming from JSON topic. Application subscribes to kafka topic inside a HostedService.  Our requirement is to have large number of kafka consumer tasks to consumer data in parallel. So we are testing with around 50  parallel tasks which are consuming from the same topic at the same time. Each  message takes around 30 to 60 milliseconds to process. All these parallel consumer tasks belong to the same consumer group id. But with this approach, we are facing memory issue where memory used by the application continues to grow linearly and at one point after couple of hours, machine gives low on virtual memory issue.  Memory used never drops. The application is running on Windows machine.  We are using Confluent.Kafka version 1.2.0.  When checked heap using VS when application was using around 4 GB of memory, around 3.58 GB memory was used by Native Heap. I am attaching the screenshot of the same. 

So my question is - Am I using Kafka consumer in a wrong way or Am I missing anything here? Is it good practice to have those many consumers(around 50) from same consumer group id consuming at the same time on windows box?

![KafkaMemoryIssue](https://user-images.githubusercontent.com/23077336/87577951-9c3b1580-c6cb-11ea-9d2f-e0a0581515af.jpg)
How to reproduce
================
// Below is the code which is mostly similar to the code we are using.
// StartAsync inside Hosted Service subscribes and creates parallel tasks 

```

 public Task StartAsync(CancellationToken cancellationToken)
        {
            var consumerConfig = new[]
            {
                new KeyValuePair<string, string>("bootstrap.servers", "enter kafka broker list here"),
                new KeyValuePair<string, string>("group.id", "enter consumer group here"),
                new KeyValuePair<string, string>("auto.offset.reset", "earliest"),
                new KeyValuePair<string, string>("enable.auto.commit", "true"),
                new KeyValuePair<string, string>("auto.commit.interval.ms", "5000")
            };
            var topicName = "enter topic name here";
            var tasks = new Task[50];
            for (var i = 0; i < 50; i++)
            {
                tasks[i] = Task.Run(async () =>
                {
                    try
                    {
                        using (IConsumer<Guid, MessageValueObject> _consumer =
                            new ConsumerBuilder<Guid, MessageValueObject>(consumerConfig)
                                  .SetKeyDeserializer(new GuidDeserializer())
                                  .SetValueDeserializer(new MessageValueDeserializer())
                                  .SetErrorHandler(ErrorHandler)
                                  .SetLogHandler(LogHandler).Build())
                        {

                            _consumer.Subscribe(topicName);
                            while (true)
                            {
                                var result = _consumer.Consume(TimeSpan.FromMilliseconds(6000));
                                if (result != null && result.Message != null)
                                {
                                    // Added below line to simulate actual processing which takes around 30 to 60 ms  
                                    await Task.Delay(30);
                                }
                            }
                        }
                    }
                    catch (ConsumeException ce)
                    {
                        throw;
                    }
                    catch (Exception ex)
                    {
                        var test3 = ex.Message;

                        throw;
                    }
                }, cancellationToken);
            }
            Task.WaitAll(tasks, cancellationToken);
            return Task.CompletedTask;
        }

```

Checklist
=========

Please provide the following information:
 - [ ] A complete (i.e. we can run it), minimal program demonstrating the problem. No need to supply a project file.
 - [ ] Confluent.Kafka nuget version.
 - [ ] Apache Kafka version.
 - [ ] Client configuration.
 - [ ] Operating system.
 - [ ] Provide logs (with "debug" : "..." as necessary in configuration).
 - [ ] Provide broker log excerpts.
 - [ ] Critical issue.

**Comment by mhowlett** [comment]
by default the consumer caches messages very aggressively in v1.4.x and below. We're changing the default settings to better match low throughput scenarios in 1.5.0, so you will probably see a big reduction just by upgrading to 1.5.0-RC1. You might also like to further reduce `QueuedMaxMessagesKbytes`, or play with related settings.

**Comment by mhowlett** [comment]
You should generally prefer to minimize the number of client instances (consumers) because it is more efficient. But using multiple consumers can sometimes make writing the code easier, and if you don't have thousands of clients, you probably won't hit any practical downside, if you tune them accordingly.

**Comment by edenhill** [comment]
Are the consumed topics used by transactional producers by any chance?

**Comment by Sandesh-J** [comment]
> Are the consumed topics used by transactional producers by any chance?

No. They are not. But I tried it by upgrading to Confluent Kafka 1.5. When my consumers were taking around  5GB of Memory, with 1.2 version, it is taking around 1 GB now with 1.5 version which is good improvement.  Also lowering value of queued.max.messages.kbytes as suggested earlier has helped with 1.2 version. 

Thanks for help!!
